thirdeye-persistence/src/main/java/ai/startree/thirdeye/datalayer/dao/TaskDao.java

```diff
@@ -264,6 +290,42 @@ public List<TaskDTO> executeParameterizedSQL(final String parameterizedSQL,
     }
   }
 
+  public TaskDTO acquireNextTaskToRun(final long workerId) {
+    try {
+      return databaseClient.executeTransaction(
+          connection -> {
+            try (final Statement s = connection.createStatement();
+                final ResultSet rs = s.executeQuery(SELECT_AND_LOCK_NEXT_TASK_QUERY)
+            ) {
+              final List<TaskEntity> res = genericResultSetMapper.mapAll(rs, TaskEntity.class);
+              if (res.size() == 1) {
+                final List<TaskDTO> dtos = toDto(res);
+                final TaskDTO toUpdate = dtos.get(0);
+                toUpdate.setStatus(TaskStatus.RUNNING);
+                toUpdate.setWorkerId(workerId);
+                toUpdate.setStartTime(System.currentTimeMillis());
+                toUpdate.setVersion(toUpdate.getVersion() + 1);
+                final int success = databaseOrm.update(toEntity(toUpdate), null, connection);
+                if (success == 1) {
+                  return toUpdate;
+                } else {
+                  throw new RuntimeException("Failed to acquire task. Failed to update the task, even though it is locked by this SQL transaction. Please reach out to StarTree support. Task id: " + toUpdate.getId());
+                }
+              } else if (res.isEmpty()) {
+                // no task to run
+                return null;
+              } else {
+                throw new RuntimeException(
+                    "Failed to acquire task. Query returned multiple rows. Only one row was expected. Please reach out to StarTree support.");
+              }
+            }
+          }, null);
+    } catch (SQLException e) {
+      LOG.error(e.getMessage(), e);
+      return null;
```

Comment:
shouldn't we raise an exception here?
maybe that's why waitForTask() is not receiving any errors in exception block

Reply:
yes we may want to throw
the real confusion comes from line 322. 
databaseClient.executeTransaction already catches all issues, except issues that happen when trying to get a Connection object.  

--> so in effect it's almost impossible to get into this catch clause.

I will do this fix in another PR. I started it but realized it was too big and impacted many other places so I removed it from this PR.
IMO it's fine if the databaseClient.executeTransaction throws exception for any error, and then it really depends on the consumer whether they want to catch the exception or not. 
The fact that databaseClient.executeTransaction does not even catch all exceptions means the callers still have to add a try catch, which results in duplication of default values and very messy code.